### ISSUE TITLE
POSTing some xml to a random endpoint crashes the API

### ISSUE THREAD

**Issue by ngaruko** [medic/cht-core#7056]
**Describe the bug**

**To Reproduce**
1. Set up a local dev deployment as [in docs](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md)
2. POST an xml form to a non-existent endpoint like , `http://localhost:5988/api/v1/forms/something`
3. Form used for this case is [death_report](https://github.com/medic/cht-core/blob/master/config/default/forms/app/death_report.xml)
4. See 404 response with 
```
{
    "error": "not_found",
    "reason": "Database does not exist."
}
```

**Expected behavior**
Expecting 404 response as above BUT Api should NOT crash
**Logs**
```
REQ 1cb478c3-8aef-4632-94dc-db42b6b0b5fc ::1 admin POST /api/v1/forms/fakepoint HTTP/1.1
RES 1cb478c3-8aef-4632-94dc-db42b6b0b5fc ::1 admin POST /api/v1/forms/fakepoint HTTP/1.1 404 58 60.366 ms
2021-04-08 01:10:19 ERROR: Server error: '{"errno":-54,"code":"ECONNRESET","syscall":"read"}' 
2021-04-08 01:10:19 ERROR: UNCAUGHT EXCEPTION! 
2021-04-08 01:10:19 ERROR:   Error: Error [ERR_STREAM_WRITE_AFTER_END]: write after end
    at writeAfterEnd (_http_outgoing.js:668:15)
    at ServerResponse.end (_http_outgoing.js:789:7)
    at ServerResponse.end (/Users/bede/tests/cht-core/api/node_modules/compression/index.js:107:21)
    at respond (/Users/bede/tests/cht-core/api/src/server-utils.js:33:7)
    at Object.serverError (/Users/bede/tests/cht-core/api/src/server-utils.js:104:5)
    at ProxyServer.<anonymous> (/Users/bede/tests/cht-core/api/src/routing.js:720:15)
    at ProxyServer.emit (/Users/bede/tests/cht-core/api/node_modules/eventemitter3/index.js:210:27)
    at ClientRequest.proxyError (/Users/bede/tests/cht-core/api/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js:165:18)
    at ClientRequest.emit (events.js:315:20)
    at Socket.socketErrorListener (_http_client.js:469:9) {
  [stack]: 'Error [ERR_STREAM_WRITE_AFTER_END]: write after end\n' +
    '    at writeAfterEnd (_http_outgoing.js:668:15)\n' +
    '    at ServerResponse.end (_http_outgoing.js:789:7)\n' +
    '    at ServerResponse.end (/Users/bede/tests/cht-core/api/node_modules/compression/index.js:107:21)\n' +
    '    at respond (/Users/bede/tests/cht-core/api/src/server-utils.js:33:7)\n' +
    '    at Object.serverError (/Users/bede/tests/cht-core/api/src/server-utils.js:104:5)\n' +
    '    at ProxyServer.<anonymous> (/Users/bede/tests/cht-core/api/src/routing.js:720:15)\n' +
    '    at ProxyServer.emit (/Users/bede/tests/cht-core/api/node_modules/eventemitter3/index.js:210:27)\n' +
    '    at ClientRequest.proxyError (/Users/bede/tests/cht-core/api/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js:165:18)\n' +
    '    at ClientRequest.emit (events.js:315:20)\n' +
    '    at Socket.socketErrorListener (_http_client.js:469:9)',
  [message]: 'write after end',
  code: 'ERR_STREAM_WRITE_AFTER_END',
  [Symbol(kEnhanceStackBeforeInspector)]: [Function: bound enhanceStackTrace] {
    [length]: 0,
    [name]: 'bound enhanceStackTrace'
  }
} 
[nodemon] app crashed - waiting for file changes before starting..
```
**Screenshots**
None
**Environment**
- Instance: local dev
- Browser: none
- Client platform:  MacOS
- App: api
- Version: 3.11

**Additional context**
Node version v14.16.0

**Comment by dianabarsan** [comment]
I'm unable to replicate this. Could you please include the whole POST, including body and headers?

**Comment by ngaruko** [comment]
I was able to reproduce it multiple times on `master` and @dianabarsan reproduced it on the `331-validate-xforms` branch. One important consideration for my environment is that I was using CouchDb desktop installation instead of the suggested [docker install](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md#couchdb-on-docker). Also interesting was that using `curl` could not crash the API , but using other clients (node code, postman and some online API testing tools like https://reqbin.com/) would break the API.

**Comment by dianabarsan** [comment]
I just saw this happen randomly on my local install, without posting XML to non existing endpoints:

![image](https://user-images.githubusercontent.com/35681649/114871689-37e7e680-9e02-11eb-83c8-82d8ac5dc9ff.png)

**Comment by dianabarsan** [comment]
It just happened again: 
![image](https://user-images.githubusercontent.com/35681649/114874023-92824200-9e04-11eb-8aa6-fea47ec4fd21.png)

**Comment by dianabarsan** [comment]
What I think is happening is caused by us proxying stuff to CouchDB when no other API endpoints "match" the request. 
When the request has a body that is large enough, and Couch processes the request without waiting for the body to be entirely received and responds with something (in the cases above, a 404). 

We intercept the error here: 
https://github.com/medic/cht-core/blob/master/api/src/routing.js#L727

and try to add our headers and content on top of what Couch responded: 
https://github.com/medic/cht-core/blob/master/api/src/server-utils.js#L33

It's probably some race condition in some node internals, since it only happens on POST requests (where something could still be streaming data) and not on GET requests for example.

**Comment by mrsarm** [comment]
Tried today again to reproduce the error without success. I tried as suggested with large XML files but the closer I get is this error:

```
REQ 502c9ed7-e6b2-4a3c-8f56-8d760ba634c7 ::ffff:127.0.0.1 admin POST /api/v1/forms/DONTEXIST HTTP/1.1
2021-06-15 18:01:49 ERROR: Server error: '{"errno":"EPIPE","code":"EPIPE","syscall":"write"}' 
RES 502c9ed7-e6b2-4a3c-8f56-8d760ba634c7 ::ffff:127.0.0.1 admin POST /api/v1/forms/DONTEXIST HTTP/1.1 500 - 5.368 ms
```

And the error above don't crash the API, and the client gets an HTTP 500 response as expected for these cases, but is a similar issue so must be related with the connection as Diana suggest.

I also tried slowing down the request so the large request body takes time to arrive completely , but that actually cause the error above to vanish, getting a 404 response as expected. I also did the requests directly to the API port, with a Nginx reverse proxy in the middle and with Ngrok. The envs used were the `master` branch, with Node v12.22.1 and with v14.16.0.

I'm wondering if the CouchDB version used affects the result, specially if the connection error happens between the API and Couch and not between the API and the user connection. I have Couch 2.3.1, what versions are you using @ngaruko and @dianabarsan ?
